projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -0,0 +1,76 @@
+import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+
+export class GraphLayout {
+
+  private levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+  private nodeToLevelMap = new Map<RenderableTopologyNode, number>();
+  private nodeCoOrdinatesMap: Map<TopologyNode, TopologyCoordinates> = new Map<TopologyNode, TopologyCoordinates>();
+
+  public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
+
+    this.findRootNodes(topology);
+    this.fillNodeAndLevelMaps();
+    this.fillNodeCoOrdinatesMap();
+    this.setNodeCoOrdinates(topology);
+  }
+
+  private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>) {
+    topology.nodes.forEach(node => {
+      if (node.incoming.length === 0) {
+        this.nodeToLevelMap.set(node, 0);
+        this.levelToNodesMap.has(0) ? this.levelToNodesMap.get(0)?.push(node) : this.levelToNodesMap.set(0, [node])
+      }
+    });
+  }
+
+  private fillNodeAndLevelMaps() {
+
+    const nodes = [...(this.levelToNodesMap.get(0) ?? [])];
+    const visited = new Set<TopologyNode>([...nodes.map(node => node.userNode)]);
+
+    while (nodes.length > 0) {
+      const parent = nodes[0];
+      const level = this.nodeToLevelMap.get(parent)! + 1;
+
+      parent.outgoing.forEach(edge => {
+        const child = edge.target;
+        if (!visited.has(child.userNode)) {
+          visited.add(child.userNode);
+          nodes.push(child);
+          this.nodeToLevelMap.set(child, level);
+          this.levelToNodesMap.has(level) ? this.levelToNodesMap.get(level)?.push(child) : this.levelToNodesMap.set(level, [child]);
```

Comment:
same point as above on root. Alternatively can pull it into a function (E.g. addToLevel)

Reply:
here if we want to intialize level to nodes map, we'll need to compute the no. of levels in advance, that I think can be a overkill

Reply:
correct - however we can do it where we assign level, rather than checking for each edge (in this approach, be careful because this may not be the first time we visited the level depending on algo used).

Reply:
Oh yeah, that will also improve the time, updated